dbms/src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -1674,8 +1727,21 @@ void MergeTreeData::removePartContributionToColumnSizes(const DataPartPtr & part
 }
 
 
-void MergeTreeData::freezePartition(const std::string & prefix, const String & with_name)
+void MergeTreeData::freezePartition(const ASTPtr & partition_ast, const String & with_name, const Context & context)
 {
+    String prefix;
+    if (format_version < MERGE_TREE_DATA_MIN_FORMAT_VERSION_WITH_CUSTOM_PARTITIONING)
+    {
+        const auto & partition = dynamic_cast<const ASTPartition &>(*partition_ast);
+        /// Month-partitioning specific - partition ID can be passed in the partition value.
+        if (const auto * partition_lit = dynamic_cast<const ASTLiteral *>(partition.value.get()))
+            prefix = partition_lit->value.getType() == Field::Types::UInt64
+                ? toString(partition_lit->value.get<UInt64>())
+                : partition_lit->value.safeGet<String>();
+    }
+    else
+        prefix = getPartitionIDFromQuery(partition_ast, context);
```

Comment:
Possibly unintended behavior if you have numeric partition key and partitions 

```
1
11
111
```

(that have common prefix)

Reply:
I am unsure if this is needed (in case of custom partitioning we can forbid prefixes and allow only exact matches), though when partitioning by date it will be handy. Proper way is of course to implement some form of ALTER PARTITION WHERE.

Reply:
On the second thought I am inclined to forbid it in case of custom partitioning. Because otherwise a user will be unable to freeze only partition '1' without also freezing partition '11'.